File: algorithm/src/main/java/baekjoon/p10819/Main.java
package baekjoon.p10819;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] inputs = new int[n];

        for (int i = 0; i < inputs.length; i++) {
            inputs[i] = sc.nextInt();
        }

        System.out.println(solve(inputs));
    }

    private static int solve(int[] inputs) {
        int[] arr = inputs.clone();
        Arrays.sort(arr);

        int ret = -1;
        for (int[] e : list(arr)) {
            ret = Math.max(ret, sum(e));
        }

        return ret;
    }

    private static List<int[]> list(int[] arr) {
        List<int[]> ret = new ArrayList<>();
        int[] next = arr;

        while (next != null) {
            ret.add(next);
            next = next(next);
        }

        return ret;
    }

    private static int[] next(int[] inputs) {
        int[] arr = inputs.clone();
        int cut = cut(arr);
        if (cut == -1) {
            return null;
        }

        int min = min(arr, arr[cut]);
        swap(arr, cut, min);
        Arrays.sort(arr, cut + 1, arr.length);

        return arr;
    }

    private static int cut(int[] arr) {
        for (int i = arr.length - 2; i >= 0; i--) {
            if (arr[i] < arr[i + 1]) {
                return i;
            }
        }

        return -1;
    }

    private static int min(int[] arr, int n) {
        for (int i = arr.length - 1; i >= 0; i--) {
            if (arr[i] > n) {
                return i;
            }
        }

        return -1;
    }

    private static void swap(int[] arr, int a, int b) {
        int temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }

    private static int sum(int[] arr) {
        int ret = 0;
        for (int i = 0; i < arr.length - 1; i++) {
            ret += Math.abs(arr[i] - arr[i + 1]);
        }

        return ret;
    }

}

File: algorithm/src/main/java/baekjoon/p10819/Main.kt
package baekjoon.p10819

import kotlin.math.abs

fun main(args: Array<String>) {
    val n = readLine()!!.toInt()
    val inputs = readLine()!!.split(" ").map { it.toInt() }
    println(solve(inputs))
}

private fun solve(inputs: List<Int>): Int {
    val arr = inputs.sorted().toIntArray()
    return list(arr).map { sum(it) }.max()!!
}

private fun list(arr: IntArray): List<IntArray> {
    val ret = ArrayList<IntArray>()
    var next = arr

    while (next.isNotEmpty()) {
        ret.add(next)
        next = next(next)
    }

    return ret
}

private fun next(input: IntArray): IntArray {
    val arr = input.clone()
    val cut = (input.lastIndex - 1 downTo 0).find { arr[it] < arr[it + 1] } ?: return intArrayOf()
    val min = (input.lastIndex downTo cut + 1).find { arr[it] > arr[cut] }!!

    swap(arr, cut, min)
    arr.sort(cut + 1)

    return arr
}

private fun swap(arr: IntArray, a: Int, b: Int) {
    val temp = arr[a]
    arr[a] = arr[b]
    arr[b] = temp
}

private fun sum(arr: IntArray): Int {
    return (0 until arr.lastIndex).sumBy { abs((arr[it] - arr[it + 1])) }
}